flowpipe/plug.py

```diff
@@ -191,3 +241,154 @@ def connect(self, plug):
         if self not in plug.connections:
             plug.connections.append(self)
             plug.is_dirty = True
+
+    def serialize(self):
+        """Serialize the Plug containing all it's connections."""
+        connections = {}
+        if self.connections:
+            connections[self.connections[0].node.identifier] = self.connections[0].name
+        return {
+            'name': self.name,
+            'value': self.value if not self._sub_plugs else None,
+            'connections': connections,
+            'sub_plugs': {
+                name: sub_plug.serialize()
+                for name, sub_plug in self._sub_plugs.items()
+            }
+        }
+
+
+class SubInputPlug(IPlug):
+    """Held by a parent input plug to form a compound plug."""
+
+    def __init__(self, name, node, parent_plug, value=None):
+        """Initialize the plug.
+
+        Can be connected to an OutputPlug.
+        Args:
+            name (str): The name of the Plug has to be in the form:
+                {parent_plug.name}.{name}.
+            node (INode): The Node holding the Plug.
+            parent_plug (InputPlug): The parent plug holding this Plug.
+        """
+        super(SubInputPlug, self).__init__(name, node, (OutputPlug, SubOutputPlug))
+        self.parent_plug = parent_plug
+        self.value = value
+        self.is_dirty = True
+
+    @property
+    def is_dirty(self):
+        """Access to the dirty status on this Plug."""
+        return self._is_dirty
+
+    @is_dirty.setter
+    def is_dirty(self, status):
```

Comment:
Question:
I keep stumbling on the handling of the dirty status. How will the parent node realize if it is clean again?

Reply:
Yep, the dirty state also gets me every time. I haven't had a use case for it yet, but with concurrency and for more user-interactive use cases, it would be crucial to have this working as expected. I'll look into this to verify it works